src/librustc_middle/mir/interpret/error.rs

```diff
@@ -148,7 +139,9 @@ impl<'tcx> ConstEvalErr<'tcx> {
             err_inval!(Layout(LayoutError::Unknown(_))) | err_inval!(TooGeneric) => {
                 return Err(ErrorHandled::TooGeneric);
             }
-            err_inval!(TypeckError) => return Err(ErrorHandled::Reported),
+            err_inval!(TypeckError(error_reported)) => {
+                return Err(ErrorHandled::Reported(error_reported));
+            }
             // We must *always* hard error on these, even if the caller wants just a lint.
             err_inval!(Layout(LayoutError::SizeOverflow(_))) => true,
```

Comment:
Hmm, this case doesn't early return so it results in `Ok(_)` and then `report_as_lint` will turn it into `ErrorReported::Lint`, which seems wrong?

Reply:
cc @oli-obk @RalfJung

Reply:
This function unfortunately became a mess over time. :/ I get the feeling we are trying to shove too many different things into one code path here.

I also do not have the slightest clue what that `replace_span_with` business in `report_as_lint` is even doing.

Reply:
But `SizeOverflow` will be a hard error even when `report_as_lint` is called, so always returning `Linted` there seems wrong, superficially.

Reply:
> This function unfortunately became a mess over time. :/ I get the feeling we are trying to shove too many different things into one code path here.

My thoughts exactly.

Reply:
Yea, so the `must_error` variable should be eliminated in favour of early returning here by calling `report_as_error`

Reply:
I kept `must_error` but replaced `Result<(), ErrorHandled>` in the return type with `ErrorHandled`, let me know if that's enough.

Reply:
that works, too, lgtm

Reply:
Avoiding that boolean could have made the code cleaner... but maybe not, and anyway we can leave that to later.